Deterministically shut down all connections on ApiListener::Stop()

jschmidt-icinga opened the pull request:
# Deterministically shut down all connections on ApiListener::Stop()

Currently, while the shutdown order is determined by icinga2 object priorities, the connection objects on the backend can continue to have effects on the state of the program, even after seemingly having been shut down (i.e. the `Stop()` method of ApiListener has been called). In practice, the connection objects will mostly stay alive until the very end when icinga2 exits and will answer all kinds of requests until then.

To mitigate some of the effects from `ApiListener`'s associated objects after `Stop()` has returned, with this PR, these additional steps will be performed (in order):

+ Exit from `ListenerCoroutineProc`
+ Disconnect remaining endpoint and anonymous `JsonRpcConnections`
+ Abort HTTP API requests that are modifying the program state and reply with `503: Service unavailable`

This PR relates to #10179 and implements one of the TODOs in @Al2Klimov's [comment](https://github.com/Icinga/icinga2/issues/10179#issuecomment-2751552284).

## Exiting from `ListnerCoroutineProc`

`ListenerCoroutineProc` is stopped by closing the acceptor socket, then waiting for the completion of outstanding `NewClientHandler`s using a separate wait group.

## Shutting down the remaining endpoint and anonymous JsonRpcConnections

Since the ListenerCoroutineProc will not take any new connections now, and `m_ReconnectTimer` has already been stopped, we can be sure that no endpoints are currently connecting and we can simply iterate over all endpoint objects and `Disconnect()` those that are still connected.

This means that some check results are still discarded when the wait group is stopped. A solution to this will require a message to be passed between the nodes and could be addressed in a future PR.

## Aborting HTTP API requests

In API requests that modify or have the potential to modify the program state, the wait group will be locked and in case of potentially longer running requests, the stopped condition will be used to abort loops that iterate over config objects early and return with a 503 error code:

```
{
    "error": 503,
    "status": "Shutting down."
}
```

fixes #10179

Al2Klimov commented on lib/checker/checkercomponent.cpp:
```diff
@@ -81,6 +81,7 @@ void CheckerComponent::Stop(bool runtimeRemoved)
 	}
 
+	m_WaitGroup->Cancel();
 	m_WaitGroup->Join();
```
You adjusted only 3/4 components, what about livestatus?

Al2Klimov commented on lib/remote/httpserverconnection.cpp:
```diff
@@ -133,7 +132,7 @@ bool EnsureValidHeaders(
 	boost::beast::http::parser<true, boost::beast::http::string_body>& parser,
 	boost::beast::http::response<boost::beast::http::string_body>& response,
-	bool& shuttingDown,
+	std::atomic<bool>& shuttingDown,
```
Please use our Atomic<>.

Al2Klimov commented on lib/remote/httpserverconnection.cpp:
```diff
@@ -71,31 +72,29 @@ void HttpServerConnection::Start()
+ * @param force ForceDisconnect() instead of GracefulDisconnect()
  */
-void HttpServerConnection::Disconnect(boost::asio::yield_context yc)
+void HttpServerConnection::Disconnect(const boost::asio::deadline_timer::duration_type & timeout)
```
So you need the new method w/o yc to call it from external?

Al2Klimov commented on lib/remote/httpserverconnection.cpp:
```diff
@@ -71,31 +72,29 @@ void HttpServerConnection::Start()
- * @param yc boost::asio::yield_context The coroutine yield context which you are calling this method from.
+ * @param force ForceDisconnect() instead of GracefulDisconnect()
  */
-void HttpServerConnection::Disconnect(boost::asio::yield_context yc)
```
But can't you keep the old one...

Al2Klimov commented on lib/remote/httpserverconnection.cpp:
```diff
@@ -71,31 +72,29 @@ void HttpServerConnection::Start()
 
-	Log(LogInformation, "HttpServerConnection")
-		<< "HTTP client disconnected (from " << m_PeerAddress << ")";
+		IoEngine::SpawnCoroutine(m_IoStrand, [this, timeout, keepAlive](asio::yield_context yc) {
```
... and call it in this coroutine?

jschmidt-icinga replied:
I can, but I was trying to make it more closely resemble the pattern used in `JsonRpcConnection`, where it works in exactly this way.

I'm happy to change it back and add an external wrapper (it's what I had at first) if the others agree that's better.

julianbrost commented on lib/remote/apilistener.cpp:
```diff
@@ -1758,6 +1802,24 @@ std::set<JsonRpcConnection::Ptr> ApiListener::GetAnonymousClients() const
+void ApiListener::DisconnectJsonRpcConnections()
+{
+	for (auto endpoint : ConfigType::GetObjectsByType<Endpoint>()) {
+		if (endpoint->GetConnected()) {
+			for (const auto& client : endpoint->GetClients()) {
+				client->Disconnect();
+			}
+		}
```
The additional `if` check around the loop can be omitted. If `GetConnected()` returned false, `GetClients()` will return an empty set anyway.

julianbrost commented on lib/remote/actionshandler.cpp:
```diff
@@ -89,6 +90,8 @@ bool ActionsHandler::HandleRequest(
 		verbose = HttpUtility::GetLastParameter(params, "verbose");
 
 	for (ConfigObject::Ptr obj : objs) {
+		std::shared_lock wgLock(*waitGroup);
```
While this function is running, [the shared lock from `HttpServerConnection::ProcessMessages()`](https://github.com/Icinga/icinga2/blob/a5b8289532641ffd622cedea9919631887bf884e/lib/remote/httpserverconnection.cpp#L481) is still held, isn't it? So this pretty much only checks whether `waitGroup` was cancelled, the unlock operation will never unblock `Join()` as there is always a surrounding lock. Given that is done in a loop, potentially many times per request, this might actually put some contention on the underlying mutex. Providing a simple function to check whether the wait group is stopped would already reduce the number of mutex operations by half here (`~shared_lock()` is no longer necessary). Making `StoppableWaitGroup::m_Stopped` atomic should even remove the need for any mutex operation here, using a single read instead.

jschmidt-icinga replied:
>So this pretty much only checks whether waitGroup was cancelled

It also throws in that case, but yes.

>this might actually put some contention on the underlying mutex

Not one that matters though, since in short timeframes futex(2) based mutexes behave like spin-locks.

>Providing a simple function to check whether the wait group is stopped

That is what `try_lock_shared()` does. It doesn't do anything else in the case the wait group is already stopped and just increment another variable if it isn't.

Sure, I could replace it with something like

``` C++
if(waitGroup.Stopped()){
    throw(StoppableWaitGroup::Stopped{});
}
```
if that's what we want.

>Making StoppableWaitGroup::m_Stopped atomic should even remove the need for any mutex operation here

Immediately, locking the Mutex is a single atomic operation. So instead of reading an atomic variable, you're setting an atomic variable, then read a regular variable and potentially set another. Yes you could make m_Stopped an atomic variable and add a getter for it, but you'd still need the mutex in the locking interface functions anyway and then you'd have doubled the processor cycles in those cases (since atomics are more expensive than regular variables and we seem to be micro-optimizing right now) and complicated the interface. 

So even if we decide on an additional `Stopped()`, I'd still advocate for keeping the mutex + regular variable. I'd ask you the same I asked @Al2Klimov in his PR, can you confidently answer the question of whether mutex contention is or will be an issue here with 'yes'?

julianbrost replied:
> Not one that matters though, since in short timeframes futex(2) based mutexes behave like spin-locks.

If you actually need mutual exclusion yes, but for checking whether that thing is stopped, you shouldn't need to wait for anyone.

> That is what `try_lock_shared()` does. It doesn't do anything else in the case the wait group is already stopped and just increment another variable if it isn't.

That's not the entire truth here. In case the try lock was successful, it also forces you to perform the reverse action (unlock).

> can you confidently answer the question of whether mutex contention is or will be an issue here with 'yes'?

The way this PR currently is, it will lock a single mutex potentially thousands[^1] of times for a single HTTP request, so we should think about whether this is a good idea.

[^1]: or even hundred of thousands of times if you're determined to do so

jschmidt-icinga replied:
>The way this PR currently is, it will lock a single mutex potentially thousands[1](https://github.com/Icinga/icinga2/pull/10460#user-content-fn-1-4ab3e006f0df7052d5853ab95f05a262) of times for a single HTTP request

I still fail to see the issue with that particular point. When we've established that locking a mutex is an atomic operation and it doesn't immediately perform a context switch or anything else expensive, how does doing a different operation on a different atomic variable thousands of times per request significantly improve things?

To me this seems more like you have an issue with the entire notion of performing this kind of repeated check. I don't like it either, maybe the alternative below will make us both happy.

>That's not the entire truth here. In case the try lock was successful, it also forces you to perform the reverse action (unlock).

Ok, I do agree with that. And also that locking the wait group isn't semantically correct in this context, even if there wasn't that locks counter that is unnecessarily incremented.

The question is what to do about it. After giving it some thought, a getter isn't as elegant a solution as it seems, since we're passing a `WaitGroup::Ptr`, not a `StoppableWaitGroup::Ptr` and I'd prefer to not move in that direction.

Maybe a better option would be passing a reference to another boolean (say an `m_Abort`; doesn't even have to be atomic) from `HttpServerConnection` down to the handlers. That would also remove the need to have a different wait group for HTTP connections because I could call the `AbortHandler()` (or whatever) in advance, then call `Join()` on the regular `m_WaitGroup` in `ApiListener`. Would that be more acceptable to you?

julianbrost replied:
> To me this seems more like you have an issue with the entire notion of performing this kind of repeated check. I don't like it either, maybe the alternative below will make us both happy.

Yes, this operation is done often enough that I'm wondering whether this is a situation where we should think about whether contention on that mutex could become a problem with parallel requests that heavily use it.

> how does doing a different operation on a different atomic variable thousands of times per request significantly improve things?

It would always be just a single atomic read. There would be no slow case under contention where it would spin on the atomic, turning the single read into many. I agree that without contention, there won't be much of a difference, the thing is that I think that would be a trivial change (i.e. expose a single bool as an atomic, no complex CAS logic needed or something like that) that would completely avoid the need about thinking if contention on a heavily used mutex might actually become a problem.

> And also that locking the wait group isn't semantically correct in this context

Why? It's unnecessary as it's enclosed by another lock so it's unnecessary, but that doesn't make it incorrect. For the typical RW lock, recursively acquiring shared locks is typically incorrect[^1], but with our wait group implementation, we know what is does and that it's fine.

[^1]: In fact, it's even [undefined behavior for `std::shared_mutex`](https://en.cppreference.com/w/cpp/thread/shared_mutex/lock_shared.html): "If lock_shared is called by a thread that already owns the mutex in any mode (exclusive or shared), the behavior is undefined"

> reference to another boolean (say an `m_Abort`; doesn't even have to be atomic)

How would you do that without atomic and without (technically) introducing a data race?

> That would also remove the need to have a different wait group for HTTP connections because I could call the `AbortHandler()` (or whatever) in advance, then call `Join()` on the regular `m_WaitGroup` in `ApiListener`. Would that be more acceptable to you?

I don't think I've fully grasped what which wait group does at the moment (see also https://github.com/Icinga/icinga2/pull/10460#discussion_r2126512015), are you saying that would allow you to reduce this back to a single wait group again?

Conceptually, this is very similar to what I would have done, though I'd have kept it inside the wait group. With your suggestion this seems to get even closer to [Go's `context` package](https://pkg.go.dev/context) and I don't have an issue with that. My only concern is that the HTTP handler function signature is getting worse and worse (see also #10142).

jschmidt-icinga replied:
>Why? It's unnecessary as it's enclosed by another lock so it's unnecessary, but that doesn't make it incorrect.

As you said, in these loops nobody needs to wait on anybody. All we want to do is signal a stop condition. In case I wasn't clear I didn't mean get rid of the outermost lock on the wait group, just the stop condition in the handler.

If there was actually something inside the handler we *wanted* ensure the completion of, once it's started, then I'd agree, but all we do is wait for a stop condition, then throw everything away we've already done and do something else.

>How would you do that without atomic and without (technically) introducing a data race?

At that point I was considering yielding the coroutine inside the handler (since we pass the yield context) and setting the boolean in `HttpServerConnection` in yet another outside accessible coroutine lambda method. But now I think yielding the coroutine, say on every iteration of the loops I currently lock the wait group in, would be orders of magnitude worse than just polling an atomic bool.

>are you saying that would allow you to reduce this back to a single wait group again?

For now it would save us the m_HttpWaitGroup, yes, because signalling the handler abort would be an extra step.

The Listener's wait group is another matter, which I have to think about some more when I've sorted out the socket close.

>My only concern is that the HTTP handler function signature is getting worse and worse (see also https://github.com/Icinga/icinga2/issues/10142).

I'm not saying we should pass it additionally, but instead. But yeah, it's still another argument.

Also, while using a struct argument is perfectly valid, I'm just going to quickly shill a std:variant + std::visit solution again here. :stuck_out_tongue_closed_eyes: I don't think the handlers need to be objects anyway, aside from needing to call virtual methods. We could register their function pointers as a variant object and then dispatch the methods with only the arguments it needs. (No need to discuss this as part of this PR though.)

julianbrost replied:
> I'm just going to quickly shill a std:variant + std::visit solution again here. [...] (No need to discuss this as part of this PR though.)

Feel free to suggest that in the issue over there though. At least I don't immediately see how you'd want to do this.

julianbrost commented on lib/remote/eventshandler.cpp:
```diff
@@ -113,6 +114,8 @@ bool EventsHandler::HandleRequest(
 	asio::const_buffer newLine ("\n", 1);
 
 	for (;;) {
+		std::shared_lock wgLock(*waitGroup);
+
 		auto event (subscriber.GetInbox()->Shift(yc));
```
That `Shift()` call is what's waiting for new events. How is this unblocked on shutdown?

jschmidt-icinga replied:
Coincidentally the timer currently waits 5s which is the same time I've set until ForceShutdown() is called on a remaining http connection (3+2s), which is why I didn't notice it sitting there for an unexpected amount of time.

But I agree this should be handled better without waiting for the timeout. I'll look into it.

julianbrost commented on lib/remote/objectqueryhandler.cpp:
I'd claim that it's enough if `waitGroup` is checked for being cancelled a few times per second. Are there any particularly expensive operations done here that justify checking it even more than once per object?

jschmidt-icinga replied:
To be honest, I'm not entirely sure. It seemed to me that depending on the request this could potentially blow up the processing time depending on what was sent in "joins". But maybe it's not needed. I'm happy to remove the inner lock in that case.

jschmidt-icinga commented on lib/base/wait-group.cpp:
```diff
@@ -27,12 +34,21 @@ void StoppableWaitGroup::unlock_shared()
+ * Wait for all existing locks to release.
  */
 void StoppableWaitGroup::Join()
 {
 	std::unique_lock lock (m_Mutex);
 
-	m_Stopped = true;
 	m_CV.wait(lock, [this] { return !m_SharedLocks; });
 }
```
Yes, that's entire point why I added the Cancel() function in the first place. If we don't want to throw away check-results that a JsonRpc node has forwarded to us already, then we need to wait on the WaitGroup without cancelling it first. But if we then want to Cancel() it with a timeout, like I do later on in ab983a3, there is no way to do that if the stopping and waiting are the same operation.

jschmidt-icinga commented on lib/remote/apilistener.cpp:
```diff
@@ -543,6 +582,11 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
 					}
 				);
 
+				std::shared_lock wgLock(*m_ListenerWaitGroup, std::try_to_lock);
+				if (!wgLock) {
+					return;
+				}
+
 				NewClientHandler(yc, strand, sslConn, String(), RoleServer);
```
Yes, the intention is to wait for it to complete if we've already started connecting to something, mostly to avoid missing any connecting clients while we Disconnect the JsonRpc clients.

julianbrost commented on lib/remote/apilistener.cpp:
```diff
@@ -512,6 +546,11 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
 
 			server->async_accept(socket.lowest_layer(), yc);
 
+			std::shared_lock wgLock(*m_ListenerWaitGroup, std::try_to_lock);
+			if (!wgLock) {
+				break;
+			}
```
> what if we've already moved the socket into the client handler coroutine and try to close the moved-from socket object?

There are two sockets in this function, `server` (the listening socket) and `socket` (the socket for an individual connection) and I was only talking about `server`, that's never moved anywhere.

julianbrost commented on lib/remote/apilistener.cpp:
```diff
@@ -368,7 +368,41 @@ void ApiListener::Stop(bool runtimeDeleted)
+	}
+
+	m_Timer->OnTimerExpired.disconnect_all_slots();
+	m_Timer->OnTimerExpired.connect([this](const Timer *const&) {
+		m_WaitGroup->Cancel();
+		DisconnectHttpClients();
+		m_Timer->Stop();
+	});
+	m_Timer->SetInterval(3);
+	m_Timer->Start();
+
+	/* HTTP API requests can take a long time to process under certain circumstances,
+	 * so unlike with check results and JSON RPC connections we want to cancel the
+	 * WorkGroup right away so it still has time to send out the 503 responses before
+	 * the AsioTlsStream timer closes the socket.
+	 */
+	m_HttpWaitGroup->Cancel();
+	m_HttpWaitGroup->Join();
+
+	/* At this point, everything is disconnected and no new connections,
+	 * JSON or HTTP API calls, check-results etc. should arrive or enter processing.
+	 * Now we wait until everything that is already being processed is done
+	 * or the timeout started above fires.
+	 */
 	m_WaitGroup->Join();
```
Isn't that more or less the situation I've described in https://github.com/Icinga/icinga2/pull/10460#discussion_r2125917636?

You do `m_WaitGroup->Cancel()` asynchronously in the background and (more or less) immediately start to `m_WaitGroup->Join()` which could return immediately if there happens to be nothing in flight at that very moment, resulting in this `Stop()` function returning. However, if there may still be messages in flight (which I assume is the case given that this 3 second window exists and `DisconnectJsonRpcConnections()` works asynchronously), if anything in there happens to do `m_WaitGroup->try_shared_lock()` after `Join()` returned but before the 3 second timer fired, `try_shared_lock()` would succeed, it would start new work, but nothing waits for it.

julianbrost commented on lib/remote/apilistener.cpp:
```diff
@@ -368,7 +368,41 @@ void ApiListener::Stop(bool runtimeDeleted)
+			<< m_HttpClients.size() << " HTTP connections remaining.";
+	}
+
+	m_Timer->OnTimerExpired.disconnect_all_slots();
+	m_Timer->OnTimerExpired.connect([this](const Timer *const&) {
+		m_WaitGroup->Cancel();
+		DisconnectHttpClients();
+		m_Timer->Stop();
+	});
+	m_Timer->SetInterval(3);
+	m_Timer->Start();
```
I'm not the biggest fan of recycling `m_Timer` here. That name already isn't very helpful and this additional usage won't make it better for anyone trying to figure out what it does.

Al2Klimov commented on lib/remote/apilistener.cpp:
```diff
@@ -368,7 +368,41 @@ void ApiListener::Stop(bool runtimeDeleted)
+	m_Timer->OnTimerExpired.disconnect_all_slots();
+	m_Timer->OnTimerExpired.connect([this](const Timer *const&) {
+		m_WaitGroup->Cancel();
+		DisconnectHttpClients();
```
Also, why to disconnect them at all? IMAO they may GET stuff as long as the process is still alive.

jschmidt-icinga commented:
No need to review this now. I'm still looking at a few things. I will update the PR description and add a comment here when I'm done.

julianbrost commented on lib/remote/apilistener.cpp:
```diff
@@ -488,11 +494,32 @@ bool ApiListener::AddListener(const String& node, const String& service)
 
 	IoEngine::SpawnCoroutine(io, [this, acceptor](asio::yield_context yc) { ListenerCoroutineProc(yc, acceptor); });
 
+	IoEngine::SpawnCoroutine(io, [this, acceptor](asio::yield_context yc) {
+		m_ShutdownListenerEvent.Wait(yc);
+		acceptor->close();
+	});
+
 	UpdateStatusFile(localEndpoint);
 
 	return true;
 }
 
+/**
+ * Stops the listener(s).
+ */
+void ApiListener::StopListener()
+{
+	namespace asio = boost::asio;
+
+	auto& io (IoEngine::Get().GetIoContext());
+
+	IoEngine::SpawnCoroutine(io, [this](asio::yield_context yc) {
+		m_ShutdownListenerEvent.Set();
+	});
+
+	m_ListenerWaitGroup->Join();
+}
```
I don't see much of a benefit in spawning two coroutines for this where the first one idles until it's woken up when the second one is started. Why not just do the work directly in the one started when the shutdown is triggered?

Apart from that, this shouldn't require spawning a full-blown stackful coroutine at all, instead, a simple callback can also be executed directly, see this for example: https://github.com/Icinga/icinga2/blob/bcf43208c64d3900d56b2c97efc0e84114cfa80a/lib/remote/jsonrpcconnection.cpp#L215

Finally, this probably needs the introduction of a strand for the listener to ensure proper synchronization: Both [`deadline_timer`](https://www.boost.org/doc/libs/1_88_0/doc/html/boost_asio/reference/deadline_timer.html#boost_asio.reference.deadline_timer.thread_safety) (used in `AsioEvent`) and [`socket`](https://www.boost.org/doc/libs/1_88_0/doc/html/boost_asio/reference/ip__tcp/socket.html#boost_asio.reference.ip__tcp.socket.thread_safety) document "Thread Safety: Shared objects: Unsafe.").

jschmidt-icinga replied:
The thread-saftety concern regarding deadline_timer was the original intention why I added the third coroutine instead of setting the `AsioEvent` directly, forgetting that they also need a strand for that to work.

And I think that `AddListener` is currently only used for a single instance, but it clearly was intended at some point to instantiate acceptors at more than one port. The other alternative to a third coroutine would be to make the acceptor object a member instead of a Shared ptr, but that is slightly past the scope of this PR.

For now I will keep the second coroutine and turn the one in `StopListener()` into a simple `post()` on the new IO strand.

julianbrost commented on lib/remote/apilistener.cpp:
```diff
@@ -1782,6 +1837,33 @@ std::set<HttpServerConnection::Ptr> ApiListener::GetHttpClients() const
+	}
+
+	Timer::Ptr gracePeriodTimer = Timer::Create();
+	gracePeriodTimer->OnTimerExpired.connect([this, gracePeriodTimer](const Timer *const&) {
```
I think you want to capture `this` as `ApiListener::Ptr` instead. The timer may outlive `DisconnectHttpClients()` and if there isn't anything waiting for it (which I don't immediately see), this could potentially access `this` after it was destroyed.

jschmidt-icinga replied:
Considering that the only `ApiListener` instance is currently static, I don't think that will happen, since the Timer thread ends long before that point.

There is still the issue that the timer object essentially lives until program end unnecessarily, since it is captured with a circular reference in the closure. I will improve this.

jschmidt-icinga commented:
Changes:
+ `ListenerCoroutineProc` is now exited by closing the socket. The wait group is still needed to track ongoing `NewClientHandler()` instances to be sure we hit all live connections in the following disconnects.
+ JsonRpcConnection::Disconnect() is still called immediately, but the lock on the wait group is now across the read and write coroutines outside the loop.
+ I've reduced the messages about disconnecting API clients to LogInformation. If we want LogWarning for the case this happens outside a normal shutdown, I can make them ask ApiListener first if this is a shutdown.
+ HttpHandlers now get passed an atomic boolean to signal the abort condition instead of the wait group. I've also reduced the situations where this abort is done as per the reviewer's suggestions. This is triggered in a seperate `AbortProcessingResponse()` method.
+ HttpServerConnection is now part of the same wait group as everything else except the one around `NewClientHandler`
+ I've moved the grace period timer inside `DisconnectHttpConnections()` and renamed them.
+ This PR now no longer adds a `Cancel()` method and the `lock_shared()` interface to `WaitGroup` since it is no longer needed (at the moment).

Caveats:
+ There is still a small potential for a situation where a connection is accepted in `ListenerCoroutineProc` and immediately closed, but I don't think it will happen often and should be harmless.
+ There is still a `m_ListenerWaitGroup`, and I'm afraid it's needed to wait on potentially multiple outstanding `NewClientHandler`s to complete. I could move this waitgroup to be local to `AddListener()` and be captured by the coroutines, but that would be more effort than I'd think is justified.
+ JsonRpcConnections may need some more attention if we want it to do what I tried to do initially, i.e. ensure that all messages other nodes have sent us will get fully processed. After thinking about this, this will at the very least involve some additional communication.

yhabteab commented on lib/remote/httpserverconnection.cpp:
```diff
@@ -122,6 +140,15 @@ void HttpServerConnection::StartStreaming()
 
+void HttpServerConnection::AbortProcessingRequest()
+{
+	if (m_HasStartedStreaming) {
+		Disconnect(boost::posix_time::seconds(0));
```
I'm not sure yet how many workarounds you'd need to achieve it, but why can't we just let events handler connections just connected as Alex already suggested? They don't trigger any kind of events but Icinga 2 just push events to them if there are any, so they wouldn't cause any problems. Though, my guess is, that would involve a lot of edge cases, so maybe at least delay their termination until the `gracePeriodTimer` from `ApiListener::DisconnectHttpClients()` is triggered.

jschmidt-icinga replied:
I could do that, sure. But in that case shutdown would always be delayed 3s whenever an event stream is active which can't be what we want, or is it?

A workaround would indeed be pretty ugly. I'd have to add a getter for `m_HasStartedStreaming` and then check it before disconnecting in the `gracePeriodTimer` callback to make an exception. Also I'd have to release the wait group somehow, no idea how to do that in a relatively sane manner.

Al2Klimov commented on lib/remote/apilistener.cpp:
```diff
@@ -486,17 +492,42 @@ bool ApiListener::AddListener(const String& node, const String& service)
+
+	IoEngine::SpawnCoroutine(m_IoStrand, [this, acceptor](asio::yield_context yc) {
+		m_ShutdownListenerEvent.Wait(yc);
+		try{
+			acceptor->close();
+		} catch (const std::exception& ex) {
+			Log(LogCritical, "ApiListener")
+				<< "Failed to close acceptor socket: " << ex.what();
+		}
```
Why can't this block...

jschmidt-icinga commented on lib/remote/apilistener.cpp:
```diff
@@ -543,8 +574,20 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
+
 				NewClientHandler(yc, strand, sslConn, String(), RoleServer);
 			});
+		} catch (const boost::system::system_error& ex) {
+			if (ex.code() == boost::system::errc::operation_canceled) {
+				break;
+			}
+
+			Log(LogCritical, "ApiListener")
+				<< "Cannot accept new connection: " << ex.what();
```
Sure, no problem.

jschmidt-icinga commented on lib/remote/apilistener.cpp:
```diff
@@ -1764,6 +1807,22 @@ std::set<JsonRpcConnection::Ptr> ApiListener::GetAnonymousClients() const
+		}
+	}
+
+	{
+		std::unique_lock lock(m_AnonymousClientsLock);
+		for (const auto & client : m_AnonymousClients){
+			client->Disconnect();
+		}
+	}
```
Nope.

jschmidt-icinga commented on lib/remote/apilistener.cpp:
```diff
@@ -1782,6 +1841,31 @@ std::set<HttpServerConnection::Ptr> ApiListener::GetHttpClients() const
+void ApiListener::DisconnectHttpClients()
+{
+	std::unique_lock lock(m_HttpClientsLock);
+	Log(LogInformation, "ApiListener")
+		<< m_HttpClients.size() << " HTTP connections remaining.";
```
Certainly. It was useful output for debugging purposes but not really needed. I'll demote it to `LogDebug` since it might be useful to know how in certain situations in conjunction with the log message that shows how many connections had to be disconnected in the end.

Al2Klimov commented on lib/remote/apilistener.cpp:
```diff
@@ -486,17 +492,42 @@ bool ApiListener::AddListener(const String& node, const String& service)
+ */
+void ApiListener::StopListener()
+{
+	boost::asio::post(m_IoStrand, [this] { m_ShutdownListenerEvent.Set(); });
```
> that would have been another refactoring making this PR larger

Not necessarily!

> I could have made the socket a member,

It's a shared pointer, so it can be additionally a member AND a local variable as-is.

jschmidt-icinga replied:
PS: I'm happy to consider at any concrete suggestions..

Al2Klimov replied:
> someone at some point must have intended a way to add multiple listeners.

Or you just use a vector of sockets as a new member.

* gets extended once in the coroutine which starts listening
* gets emptied in the callback which shuts down the listeners

jschmidt-icinga replied:
We've talked about it yesterday and decided that it is past the scope of this PR. I fully agree that this could do with a refactor in either direction. You're welcome to open a PR when/if this gets merged.

yhabteab commented on lib/remote/apilistener.cpp:
```diff
@@ -1782,6 +1841,31 @@ std::set<HttpServerConnection::Ptr> ApiListener::GetHttpClients() const
+	Log(LogInformation, "ApiListener")
+		<< m_HttpClients.size() << " HTTP connections remaining.";
+
+	for (const auto& client : m_HttpClients) {
+		client->AbortProcessingRequest();
+	}
+
+	auto gracePeriodTimer = Shared<boost::asio::deadline_timer>::Make(IoEngine::Get().GetIoContext());
+	gracePeriodTimer->expires_from_now(boost::posix_time::seconds(3));
+	gracePeriodTimer->async_wait([this, gracePeriodTimer](const boost::system::error_code & ec) {
```
You're right, the `async_wait()` shouldn't yield in that case and in its docs states:
> Regardless of whether the asynchronous operation completes immediately or not, the completion handler will not be invoked from within this function. On immediate completion, invocation of the handler will be performed in a manner equivalent to using [post](https://live.boost.org/doc/libs/1_85_0/doc/html/boost_asio/reference/post.html).

So, it shouldn't cause any problems but I would still scope the mutex lock to the range-loop only via `{...}`.

jschmidt-icinga commented:
Again I've simplified and reduced the scope of the PR, according to the suggestion of @Al2Klimov. I've also think I've addressed the remaining comments that still apply.

Al2Klimov commented on lib/remote/actionshandler.cpp:
```diff
@@ -88,7 +89,18 @@ bool ActionsHandler::HandleRequest(
+	}
+
 	for (ConfigObject::Ptr obj : objs) {
+		if (!waitGroup->IsLockable()) {
+			HttpUtility::SendJsonError(response, params, 503, "Shutting down.");
+			return true;
+		}
```
As the client, I'd also wonder if I get results only for 1/2 of the requested checkables btw. On the other hand, a total 503 like here implies nothing at all succeeded. Difficult moral question... 🙈

julianbrost replied:
I don't agree that sending a response for only half the objects would be a bad idea if that's what we processed. That doesn't get any better by hiding it. Some middle ground would be to still iterate over all objects but once the shutdown request was propagated here, just add a corresponding status message for each (similar to how errors for individual objects are already reported in the `catch` block a few lines below).

julianbrost replied:
Should be fine. You could also additionally prefix it with something like "Action skipped:" for consistency with the other string.

jschmidt-icinga commented on lib/remote/jsonrpcconnection.cpp:
```diff
@@ -68,6 +68,11 @@ void JsonRpcConnection::HandleIncomingMessages(boost::asio::yield_context yc)
 
 	m_Stream->next_layer().SetSeen(&m_Seen);
 
+	std::shared_lock wgLock(*m_WaitGroup, std::try_to_lock);
+	if (!wgLock) {
+		return;
+	}
```
From my testing, yes. In the previous iteration of the PR I was scared this would be a similar situation as the `async_write()` in `HttpServerConnection` blocking the shutdown. But unlike there, here the coroutine mostly sits in the `async_read()` without actually making progress (i.e. giving other operations a chance to run) or actively reading the rather small RPC messages (like check results).

Then again, the same corner cases could arise. Large messages over very, very slow connections. But even in that case, I think it would be better to let the message finish than to interrupt reading it in the middle, even if it takes a second or two.

To move around this, acquiring the lock somewhere else isn't even enough, I would have to pass something into `NetString::ReadStringFromStream` to interrupt reading the message, which I really don't want to do.

jschmidt-icinga replied:
>I don't see how an implementation without introducing new cluster messages [...]

I get that and also commented on it [here (last point)](https://github.com/Icinga/icinga2/pull/10460#issuecomment-2949356681).

>I'd totally fine with only counting the message processing itself, no the read operation.

Alright, so I'd move to lock to the top of `MessageHandler()`. Is that fine? Should the lock in `WriteOutgoingMessages()` stay then or should I rather remove it entirely? There's a timeout in `Disconnect()`, but I guess that can fail to cancel the operation too under certain conditions.

jschmidt-icinga commented on lib/base/wait-group.cpp:
```diff
@@ -26,13 +26,18 @@ void StoppableWaitGroup::unlock_shared()
 	std::unique_lock lock (m_Mutex);
 
-	m_Stopped = true;
+	m_Stopped.store(true, std::memory_order_relaxed);
```
At the very least I would need to use `.store()` due to the constraints of the `Atomic<bool>` class vs. `std::atomic<bool>`. And since I'm already doing that I might as well put `std::memory_order_relaxed` here.

jschmidt-icinga commented:
I've pushed and update that should address the remaining comments, with the exception of the wait group lock in `JsonRpcConnection::WriteOutgoingMessages`, which I'm still not sure about.

Another change is that I'm now using a boost::signal+post instead of the post+coroutine I used to stop the listener before. I've also moved the lock inside the loop because I had after all encountered rare situations where, like @julianbrost warned, the `operation_aborted` error was not thrown and the socket was closed after `async_accept` succeeded, trapping `ListenerCoroutineProc` in an endless loop of `bad_file_descriptor` exceptions. Asio coroutines are a lot of fun.

julianbrost commented:
> the `operation_aborted` error was not thrown and the socket was closed after `async_accept` succeeded, trapping `ListenerCoroutineProc` in an endless loop of `bad_file_descriptor` exceptions

Did this just happen randomly? Were you able to make this happen more reliably by creating some specific scenario?

> I've also moved the lock inside the loop

Does that prevent the `bad_file_descriptor` situation itself or just the endless loop as that locking operation acts pretty much like a loop condition (like for example `while (acceptor->is_open())` or `while (m_ListenerWaitGroup->IsLockable())`)?

julianbrost commented on lib/remote/consolehandler.cpp:
```diff
@@ -94,6 +95,12 @@ bool ConsoleHandler::HandleRequest(
 		return true;
 	}
 
+	std::shared_lock wgLock{*waitGroup, std::try_to_lock};
+	if (!wgLock) {
+		HttpUtility::SendJsonError(response, params, 503, "Shutting down.");
+		return true;
+	}
```
I wouldn't add the lock here, or at least not keep it while executing an unsandboxed expression. The [warnings in the documentation](https://icinga.com/docs/icinga-2/latest/doc/12-icinga2-api/#debug-console) are to be taken seriously and if you do anything where this lock might help you, you are on your own already. And unsandboxed really means that. If you submit an endless loop, that request handler will execute it and with the lock still held, that results in an Icinga 2 process that can't shut down anymore.

jschmidt-icinga replied:
Makes sense.

julianbrost commented:
Sorry, I forgot to address the following in a previous comment/review:

> with the exception of the wait group lock in JsonRpcConnection::WriteOutgoingMessages, which I'm still not sure about.

For me, it's pretty similar to the one in the read loop and don't add unless:
1. You can reason that the socket operations that are done while it's held are unblocked quickly on shutdown; or
2. You say it's needed for correctness, but in that case we need to ensure that that socket operations are unblocked at shutdown.

jschmidt-icinga commented:
>Did this just happen randomly? Were you able to make this happen more reliably by creating some specific scenario?

I think it's the same situation I've observed in #5581, with the coroutine yielding again when the async operation is actually already done. Here, this would mean closing the socket comes too late to make async_accept() throw, but obviously still closes the socket. I thought about alternatively checking if the wait group is not lockable anymore when catching the `bad_file_descriptor` exception, but I liked this option better and it should avoid the exception entirely, at least I moved the lock.

>that locking operation acts pretty much like a loop condition

Well, it acts as locking *and* a loop condition. I can also keep it outside the loop and just do `while (IsLockable())` if you prefer it that way. Wouldn't even need to check the shared_lock if we use it as a loop condition.

jschmidt-icinga commented on lib/remote/apilistener.cpp:
```diff
@@ -486,13 +491,38 @@ bool ApiListener::AddListener(const String& node, const String& service)
+	 */
+	auto strand = Shared<asio::io_context::strand>::Make(io);
+
+	m_OnListenerShutdown.connect([strand, acceptor]() {
+		boost::asio::post(*strand, [acceptor] {
+			try {
+				acceptor->close();
+			} catch (const std::exception& ex) {
+				Log(LogCritical, "ApiListener")
+					<< "Failed to close acceptor socket: " << ex.what();
+			}
+		});
+	});
```
Technically it's cleaned up when ApiListener ist destroyed unless I'm missing something. And since we only call it once (on potentially, but not really, multiple listeners) I thought it would be fine. But thanks for pointing me to `scoped_connection`, I didn't know about that and will use that.

jschmidt-icinga commented on lib/remote/actionshandler.cpp:
```diff
@@ -88,7 +89,24 @@ bool ActionsHandler::HandleRequest(
+		return true;
+	}
+
 	for (ConfigObject::Ptr obj : objs) {
+		if (!waitGroup->IsLockable()) {
+			Dictionary::Ptr fail = new Dictionary({
```
Would "early" in this case really save a lot of time justifying this?

>I'm not sure if (ab)using wgLock as a state variable for "have we given up this request" is really a good idea:

Maybe something like this:

``` c++
if (!waitGroup->IsLockable()) {
	if (wgLock.owns_lock()) {
		wgLock.unlock();
	}
	Dictionary::Ptr fail = new Dictionary({
	...
```

julianbrost replied:
> Would "early" in this case really save a lot of time justifying this?

I'm not sure about the actual impact, it was more of a general "narrow lock scope to the relevant part" idea. I mean there could be a huge number of objects left, and this will perform quite a number of memory allocations, however, the worst case probably is that if you have a huge request and particularly bad timing, your shutdown will take a bit longer, but probably not even a second longer. So treat it a a random thought of mine and feel free to ignore it.

jschmidt-icinga commented on lib/remote/deleteobjecthandler.cpp:
```diff
@@ -78,7 +79,18 @@ bool DeleteObjectHandler::HandleRequest(
+	}
+
 	for (ConfigObject::Ptr obj : objs) {
+		if (!waitGroup->IsLockable()) {
+			HttpUtility::SendJsonError(response, params, 503, "Shutting down.");
+			return true;
+		}
```
Sure, I'll add equivalent sections to these two.

jschmidt-icinga commented:
> 2\. You say it's needed for correctness, but in that case we need to ensure that that socket operations are unblocked at shutdown.

We can ensure they are unblocked, but not guarantee it's within a reasonable time, even when the 5s timeout in `Disconnect()` fires, the `async_write()` will not yield on very slow but steady connections, same as in `HttpServerConnection`. I'll just remove it.

julianbrost commented:
> the same situation I've observed in #5581

Is that really the correct PR you wanted to reference?

I was asking primarily because I didn't understand how changing the scope of the locking should solve bad file descriptor errors and if these appear randomly, that's always a hint that something could be wrong with the code.[^1]

In the current state of the PR, I think there could still be a small race condition in that respect:

https://github.com/Icinga/icinga2/blob/44c902de334d6558c0b92a5636ae9953e26fb59b/lib/remote/apilistener.cpp#L519-L524

The accepting socket is already closed by line 521, 


[^1]: If a syscall actually returns `EBADFD` that could be a hint that a fd number is used after it was closed. At least in a multi-threaded application, instead of receiving that error, there's a chance that the fd number could have been reused already and instead of getting an error, the operation could also have been performed on a random other file instead. However, with Asio in between, that also [maps operations on closed socket objects (which wrap a fd) to that error without ever issuing a syscall](https://github.com/boostorg/asio/blob/boost-1.88.0/include/boost/asio/detail/impl/socket_ops.ipp#L116-L120) so that use-after-close isn't an issue (unless you have code that reuses the Asio socket object).

jschmidt-icinga commented:
>Is that really the correct PR you wanted to reference?

Sorry, no. I screwed up the auto complete. I meant #10139.

>and if these appear randomly, that's always a hint that something could be wrong with the code.

I mean, it didn't appear *randomly*, it appeared *sporadically* when due to timing closing the socket didn't cancel an ongoing `async_accept()`.

Anyway, I've just pushed yet another update.

Note: I just noticed that `ModifyObjectsHandler` currently always sends `http::status::ok` as a result, despite other error conditions existing. I'm not going to change that here, but maybe doing the same as `DeleteObjectsHandler` would be a good thing.

julianbrost commented:
> I mean, it didn't appear _randomly_, it appeared _sporadically_

Randomly in the sense that I don't understand the pattern and it looks random to me. ;)

Just remove "randomly" from my sentence. seeing a bad file descriptor error anywhere is rarely a good sign.

julianbrost commented on lib/remote/deleteobjecthandler.cpp:
```diff
@@ -78,7 +79,29 @@ bool DeleteObjectHandler::HandleRequest(
+				wgLock.unlock();
+			}
+
+			Dictionary::Ptr fail = new Dictionary({
+				{ "code", 503 },
+				{ "status", "Action skipped: Shutting down."}
+			});
```
It would be a good idea to still include "type" and "name" like below so you at least have a chance of knowing which objects weren't deleted:

https://github.com/Icinga/icinga2/blob/fb81054ef39c7cfa27ad00979ed6ede6119ffe60/lib/remote/deleteobjecthandler.cpp#L123-L124

(Same for `ModifyObjectHandler`)

julianbrost commented on lib/remote/actionshandler.cpp:
```diff
@@ -88,7 +89,28 @@ bool ActionsHandler::HandleRequest(
+				wgLock.unlock();
+			}
+			
+			Dictionary::Ptr fail = new Dictionary({
+				{ "code", 503 },
+				{ "status", "Action skipped: Shutting down."}
+			});
+
+			results.emplace_back(std::move(fail));
```
I'd also add "type" and "name" to the response like my other comment suggests for the other two handlers. Here, the other cases don't already do this, but with the current code, you just get a bunch of repetitions of that very same JSON without any reference to the object. So repeating it doesn't really provide any benefit.

julianbrost replied:
I'd just create that dictionary in the argument of `emplace_back()`. At least I see little value in having a variable here just to move it in the next step.

julianbrost commented on lib/remote/apilistener.cpp:
```diff
@@ -486,13 +491,40 @@ bool ApiListener::AddListener(const String& node, const String& service)
 		<< "Started new listener on '[" << localEndpoint.address() << "]:" << localEndpoint.port() << "'";
 
-	IoEngine::SpawnCoroutine(io, [this, acceptor](asio::yield_context yc) { ListenerCoroutineProc(yc, acceptor); });
+	/* The Asio strand for the listener socket.
+	 */
```
```suggestion
	// The Asio strand for the listener socket.
```

Or in case you only added this comment because I asked for one documenting the scope of the strand earlier: When I asked for that, the strand was a class member with it's use spread across more functions. Now it's all done here in this function in close proximity and also the Asio shutdown signal that was additionally guarded is gone, so I'd also be fine with removing the comment.

julianbrost commented on lib/base/wait-group.cpp:
```diff
@@ -26,13 +26,18 @@ void StoppableWaitGroup::unlock_shared()
 	}
 }
 
+bool StoppableWaitGroup::IsLockable () const
```
```suggestion
bool StoppableWaitGroup::IsLockable() const
```

julianbrost commented on lib/remote/actionshandler.cpp:
```diff
@@ -88,7 +89,28 @@ bool ActionsHandler::HandleRequest(
+			if (wgLock) {
+				wgLock.unlock();
+			}
+			
```
```suggestion

```

(There is trailing whitespace in this line, I'm not sure how clear this will be from what GitHub shows, I'm **not** asking to remove the empty line.)

julianbrost commented on lib/remote/apilistener.cpp:
```diff
@@ -506,7 +538,14 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
+			<< "Could not lock the listener wait group.";
+		return;
+	}
+	
```
```suggestion

```

(Also trailing whitespace)

julianbrost reviewed (approved):
Thank you very much!

> @Al2Klimov @yhabteab I only expect some small changes here, so in case you want to take another look, now (or after the next push) would be a good time.

Not merging yet until I get some feedback there.
